dev/reloadable-server/src/main/scala/play/core/server/LagomReloadableDevServerStart.scala

```diff
@@ -132,98 +132,83 @@ object LagomReloadableDevServerStart {
 
             synchronized {
 
-              // Let's load the application on another thread
-              // as we are now on the Netty IO thread.
-              //
-              // this whole Await.result(Future{}) thing has been revisited in Play
-              // but the issue is not yet fully addressed there neither
-              // see issues:
-              // https://github.com/playframework/playframework/pull/7627
-              // and https://github.com/playframework/playframework/pull/7644
-              //
-              // we should reconsider if still need to have our own reloadable server
-              implicit val ec = ExecutionContext.global
-              Await.result(scala.concurrent.Future {
```

Comment:
(Copying comment from the previous commit)

The main change was to take out the `Future` spawn + `Await`. It's not obvious what the point of that was, but I believe that blocking on the global execution context (without signaling it using [`blocking`](https://www.scala-lang.org/api/current/scala/concurrent/index.html#blocking[T](body:=%3ET):T)) was causing the global execution context to run out of threads.

By default, the global execution context has max threads = #CPUs, making this more likely to occur in constrained environments such as Travis CI workers.

Note that, prior to the changes in #1405, this [ran on the Play internal execution context](https://github.com/lagom/lagom/pull/1405/files#diff-b0570315055c9029853b183c2943bf7aL139), which has been removed in Play 2.7.

Reply:
One clarification: `Await.result` calls `blocking` internally, but then the body of the `Future` does other blocking things without signalling.

Reply:
I think the problem is not so much caused by the `Await` as it is caused by an `Executor` with an unpredictable size (and general purpose). I think we could have an `Executor` to be used on DevMode in Lagom to avoid using the default, constrained one.

Reply:
e.g.

```scala
              import java.util.concurrent.atomic.AtomicInteger
              import java.util.concurrent.{ Executors, ThreadFactory }

              implicit val ex = ExecutionContext.fromExecutor(
                Executors.newFixedThreadPool(10, new ThreadFactory {
                  private val counter = new AtomicInteger(0)
                  override def newThread(r: Runnable): Thread = {
                    new Thread(r, s"lagom-devmode-buildlink-${counter.incrementAndGet()}")
                  }
                })
              )
```

Reply:
There's still the question: why did it work-as-expected on sbt `1.1.4` and not in sbt `1.1.5`.

Reply:
> I think the problem is not so much caused by the Await as it is caused by an Executor with an unpredictable size (and general purpose). 

Forgot to comment on this. I think you are right here. We didn't have this problem when using sbt 1.1.6 but not global execution context: https://github.com/lagom/lagom/pull/1400.

Reply:
BTW: I've run the tests successfully using the `Executor` [above](https://github.com/lagom/lagom/pull/1441#discussion_r202352420) and the original code (`Await.result(scala.concurrent.Future {...`).

Reply:
The lock involves `sbt`'s [`ConcurrentRestrictions`](https://github.com/sbt/sbt/blob/a41727fb17bb923036f90ca2ca62897def1a893e/tasks/src/main/scala/sbt/ConcurrentRestrictions.scala#L207) awaiting indefinitely for a task of a [`ExecutorCompletionService`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorCompletionService.html) to finish. I've checked and I think the `ExecutorCompletionService` is not sharing any resource with `ExecutionContext.global` so I'm not sure how the `sbt` code deadlocks with `LagomReloadableDevServerStart`.

Reply:
> I think the problem is not so much caused by the `Await` as it is caused by an `Executor` with an unpredictable size (and general purpose).

That's right, sorry that my explanation wasn't clear.

> I think we could have an `Executor` to be used on DevMode in Lagom to avoid using the default, constrained one.

We _could_, but why? That's what I was trying to convey: I don't see a reason to spawn a future at all. Is there some reason not to run that process on the current thread? I couldn't see one. Note that the code in Play this was derived from already made this change in https://github.com/playframework/playframework/pull/7627 (along with some other changes that I think aren't essential). Ultimately, we should reconcile this and use the upstream implementation in Play directly, but for the purposes of getting this working again, this seems good enough.

@richdougherty you reviewed https://github.com/playframework/playframework/pull/7627, and this relates to your comment https://github.com/playframework/playframework/pull/7627#issuecomment-315640287. What do you think about this?

> There's still the question: why did it work-as-expected on sbt 1.1.4 and not in sbt 1.1.5?

I don't have a clear answer for this, and I'm wary of diving down a rabbit hole, but from looking at the thread dump you sent I suspect it could be related to b78a0f667be2c0d27d4f313aeacb98a53833d2cd. @dwijnand what do you think?

[thread-dump.txt](https://github.com/lagom/lagom/files/2196443/thread-dump.txt)

@marcospereira you could be right that it's related to the file-watching changes, too, but I don't see anything in the thread dump that indicates that.

> The lock involves `sbt`'s [`ConcurrentRestrictions`](https://github.com/sbt/sbt/blob/a41727fb17bb923036f90ca2ca62897def1a893e/tasks/src/main/scala/sbt/ConcurrentRestrictions.scala#L207) awaiting indefinitely for a task of a [`ExecutorCompletionService`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorCompletionService.html) to finish. I've checked and I think the `ExecutorCompletionService` is not sharing any resource with `ExecutionContext.global` so I'm not sure how the `sbt` code deadlocks with `LagomReloadableDevServerStart`.

My latest theory is that the `catch` that was removed in https://github.com/sbt/sbt/commit/b78a0f667be2c0d27d4f313aeacb98a53833d2cd#diff-ca2e29d21ecb145cb799f9f722c5ececL94 was handling a `RejectedExecutionException` that was thrown from somewhere other than `CompletionService.submit`.

So, maybe changing the execution context results in a change to the execution stacks, meaning that an exception is being handled by some other catch somewhere. Or, possible the `RejectedExecutionException` is thrown when using a fork-join-pool execution context like the global one, but not when the code is executed synchronously. The docs are a little ambiguous:

> This implementation rejects submitted tasks (that is, by throwing RejectedExecutionException) only when the pool is shut down or internal resources have been exhausted.

I'm not sure what constitutes internal resources being exhausted.

I had a look at the [diff between Scala 2.12.4 and 2.12.6](https://github.com/scala/scala/compare/v2.12.4...v2.12.6), but it's huge.

Could be related to https://github.com/scala/scala/pull/6159 or https://github.com/scala/scala/pull/6171.

Reply:
Why this? : https://github.com/lagom/lagom/pull/1441/files#diff-b0570315055c9029853b183c2943bf7aR133

Reply:
You mean why is the whole thing synchronized? I'm not 100% positive, but I assume it's to avoid two concurrent reloads of the same service (`this` is one per service). @richdougherty?